src/main/java/spoon/reflect/visitor/PrinterTokenDetector.java

```diff
@@ -0,0 +1,299 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import spoon.SpoonException;
+import spoon.reflect.cu.CompilationUnit;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.visitor.printer.TokenType;
+
+/**
+ * The special {@link PrinterHelper} implementation,
+ * which detects printed tokens and sends them to registered {@link PrinterTokenWriter}
+ *
+ * Nothing is sent to parent PrinterHelper, which is extended just to assure minimal code changes, but which is not really used
+ */
+class PrinterTokenDetector extends PrinterHelper {
```

Comment:
why having a subclass? why not modifying directly PrinterHelper and add there the PrinterTokenWriter? (since PrinterTokenDetector is already hard-wired in DJPP)

Reply:
I will check what I can do here to avoid subclassing. 

Just note: I am sure that we cannot add behavior of PrinterTokenWriter into PrinterHelper, because PrinterHelper with origin behavior is needed at the end of the token chain too.